syntax = "proto3";

package protocol;

message Endpoint {
    string network = 1;
    string address = 2;
    sint64 target_id = 3;
}

message SubscribeOption {
    string service = 1;
    uint32 min_distance = 2;
    uint32 max_distance = 3;
}

message PublishSettings {
    uint32 priority = 1;
    uint32 max_distance = 2;
    uint32 concurrency = 3;
    uint32 queue_size = 4;
}

message Route {
    repeated SubscribeOption options = 1;
    uint32 weight = 2;
}

message ServiceChange {
    sint64 target_id = 1;
    string name = 2;
    PublishSettings settings = 3;
}

message SubscribeRequest {
    repeated SubscribeOption options = 1;
    bool disable_routes = 2;
}

message SubscribeResponse {
    sint64 target_id = 1;
}

message GetEndpointsRequest {
    sint64 target_id = 1;
}

message GetEndpointsResponse {
    repeated Endpoint endpoints = 1;
}

message StreamEndpointsRequest {
    sint64 target_id = 1;
}

message StreamEndpointsResponse {
    repeated Endpoint added = 1;
    repeated Endpoint removed = 2;
}

message UnsubscribeRequest {
    sint64 target_id = 1;
}

message UnsubscribeResponse {
    // nothing yet
}

message PublishRequest {
    sint64 target_id = 1;
    string name = 2;
    PublishSettings settings = 3;
}

message PublishResponse {
    // nothing yet
}

message UnpublishRequest {
    sint64 target_id = 1;
}

message UnpublishResponse {
    // nothing yet
}

message SetRouteRequest {
    string name = 1;
    repeated Route routes = 2;
}

message SetRouteResponse {
    // nothing yet
}

message ListRoutesRequest {
    // nothing yet
}

message ListRoutesResponse {
    repeated string names = 1;
}

message LookupRouteRequest {
    string name = 1;
}

message LookupRouteResponse {
    repeated Route routes = 1;
}

message StreamServicesRequest {
    // nothing yet
}

message StreamServicesResponse {
    repeated sint64 removed = 1 [packed=true];
    repeated ServiceChange changed = 2;
}

enum RequestType {
    NewStream = 0;
    Subscribe = 1;
    GetEndpoints = 2;
    StreamEndpoints = 3;
    Unsubscribe = 4;
    Publish = 5;
    Unpublish = 6;
    SetRoute = 7;
    ListRoutes = 8;
    LookupRoute = 9;
    StreamServices = 10;
}
